### ISSUE TITLE
[FEATURE] Mechanism to “pause” a useAnimationFrame

### ISSUE THREAD

**Issue by controversial** [framer/motion#2046]
### Is your feature request related to a problem? Please describe.
useAnimationFrame runs its callback on every animation frame, but in many scenarios, the loop is only necessary under some conditions that are unrelated to whether the component is mounted on the page. For example, on a long-scrolling page, it might be appropriate to use `IntersectionObserver` to decide to only run an animation when it will be visible to the user.

### Describe the solution you'd like
I’d like a way to “pause” a `useAnimationFrame` without unmounting the component—so that its render loop stops—without unmounting the component.

This could be an imperative API exposed by the `useAnimationFrame` hook:
```js
const handle = useAnimationFrame(...);
...
handle.stop();
handle.start();
```
or a declarative API as part of the useAnimationFrame call:
```js
const [shouldRun, setShouldRun] = useState(true);
useAnimationFrame(..., shouldRun);
```

### Describe alternatives you've considered
I’ve considered making my own requestAnimationFrame loop, but this throws out some of the performance tuning benefits of `framer-motion`.

**Comment by mattgperry** [comment]
Yeah good idea, I've just opened a PR

**Comment by mattgperry** [comment]
I decided against implementing this. On inspection it's possible to stop the loop simply by not passing a callback.

```
useAnimationFrame(isActive ? callback : undefined)
```

**Comment by lucienvoltzcnp** [comment]
> I decided against implementing this. On inspection it's possible to stop the loop simply by not passing a callback.
> 
> ```
> useAnimationFrame(isActive ? callback : undefined)
> ```

```
  useAnimationFrame(
    hovered
      ? (time, delta) => {
          if (!testRef.current) return;
          testRef.current.innerHTML = `rotateY(${time}deg)`;
        }
      : undefined
  );
```

This breaks when using typescript, `'((time: number, delta: number) => void) | undefined' is not assignable to parameter of type 'FrameCallback'`

**Comment by jamestrenda** [comment]
> > I decided against implementing this. On inspection it's possible to stop the loop simply by not passing a callback.
> > ```
> > useAnimationFrame(isActive ? callback : undefined)
> > ```
> 
> ```
>   useAnimationFrame(
>     hovered
>       ? (time, delta) => {
>           if (!testRef.current) return;
>           testRef.current.innerHTML = `rotateY(${time}deg)`;
>         }
>       : undefined
>   );
> ```
> 
> This breaks when using typescript, `'((time: number, delta: number) => void) | undefined' is not assignable to parameter of type 'FrameCallback'`

True, but you can do this: 

```
useAnimationFrame(
    hovered
      ? (time, delta) => {
          if (!testRef.current) return;
          testRef.current.innerHTML = `rotateY(${time}deg)`;
        }
      : () => { // do something else }
  );
```
  
But that's also not very helpful unless you can somehow track `t` from outside the hook. I got ChatGPT to help out here with a looping text animation that seems to do the trick:
  
  ```
  const [scope, animate] = useAnimate();
  const inView = useInView(scope);

  const [wordIndex, cycleWordIndex] = useCycle([...Array(words.length).keys()]);
  const wordRef = useRef(null);
  const prev = useRef(-1);

  // setup refs for the pause hack...
  const startTime = useRef<number | null>(null);
  const elapsedTime = useRef(0);

  useAnimationFrame(
    inView
      ? (t) => {
          if (!startTime.current) {
            startTime.current = t - elapsedTime.current; // Adjust start time to resume from the last elapsed time
          }

          const currentElapsedTime = t - startTime.current;
          elapsedTime.current = currentElapsedTime;
          
          
         // begin project-specific logic....

          const seconds = Math.round(currentElapsedTime / 2000);
          const trigger = seconds % words.length;

          if (trigger !== prev.current) {
            prev.current = trigger;
            cycleWordIndex();
          }

          // end project-specific logic

        }
      : () => (startTime.current = null) // reset our timer ref
  );

**Comment by FrontEndNoah** [comment]
> I decided against implementing this. On inspection it's possible to stop the loop simply by not passing a callback.
> 
> ```
> useAnimationFrame(isActive ? callback : undefined)
> ```

I don't think this will solve the problem, as it looks more like a destroy than a pause.

This solution is only valid when the default value of isActive is true, If the default value is false and then changed to true, the callback function will not be updated

I think the person who raised this question probably hopes to avoid having to record the timestamp offset himself, because it is easy to involve pauses when developing animations. Therefore, the other way of switching to an empty function is not possible, because the timestamp will continue to accumulate, and there will still be a timestamp offset after switching back, which still needs to be manually recorded and corrected.

So, to sum up, I think this feature can still be considered for implementation, otherwise developers using this hook will have to handle the offset themselves, or re-encapsulate a hook that can be paused.
